internal/app/machined/pkg/controllers/cri/image_cache_config.go

```diff
@@ -287,16 +293,116 @@ func (ctrl *ImageCacheConfigController) createVolumeConfigDisk(ctx context.Conte
+			}
+
+			return nil, fmt.Errorf("error getting volume status: %w", err)
+		}
+
+		volumeStatuses = append(volumeStatuses, volumeStatus)
+	}
+
+	// we need to ensure that we first wait for the ISO to be either missing or ready,
+	// so that we can make a decision on copying the image cache from an ISO to the disk volume
+	var isoStatus, diskStatus block.VolumePhase
+
+	for _, volumeStatus := range volumeStatuses {
+		switch volumeStatus.Metadata().ID() {
+		case VolumeImageCacheISO:
+			isoStatus = volumeStatus.TypedSpec().Phase
+		case VolumeImageCacheDISK:
+			diskStatus = volumeStatus.TypedSpec().Phase
+		}
+	}
+
+	if isoStatus != block.VolumePhaseMissing && isoStatus != block.VolumePhaseReady {
+		return &imageCacheVolumeStatus{}, nil
+	}
+
+	isoPresent := isoStatus == block.VolumePhaseReady
+	diskMissing := diskStatus == block.VolumePhaseMissing
+
+	roots := make([]string, 0, len(volumeIDs))
+
+	var (
+		allReady, isoReady, diskReady bool
+		copySource, copyTarget        string
+	)
+
+	// analyze volume statuses, and build the roots
+	for _, volumeStatus := range volumeStatuses {
+		// mount as rw only disk cache if the ISO cache is present
+		root, ready, err := ctrl.getImageCacheRoot(ctx, r, volumeStatus, !(volumeStatus.Metadata().ID() == VolumeImageCacheDISK && isoPresent))
+		if err != nil {
+			return nil, fmt.Errorf("error getting image cache root: %w", err)
+		}
+
+		if ready {
+			switch volumeStatus.Metadata().ID() {
+			case VolumeImageCacheISO:
+				isoReady = true
+				copySource = root.ValueOr("")
+			case VolumeImageCacheDISK:
+				diskReady = true
+				copyTarget = root.ValueOr("")
+			}
+		}
+
+		allReady = allReady && ready
+
+		if rootPath, ok := root.Get(); ok {
+			roots = append(roots, rootPath)
+		}
+	}
+
+	var copyStatus cri.ImageCacheCopyStatus
+
+	switch {
+	case !isoPresent:
+		// if there's no ISO, we don't need to copy anything
+		copyStatus = cri.ImageCacheCopyStatusSkipped
+	case diskMissing:
+		// if the disk volume is not configured, we can't copy the image cache
+		copyStatus = cri.ImageCacheCopyStatusSkipped
+	case ctrl.cacheCopyDone:
+		// if the copy has already been done, we don't need to do it again
+		copyStatus = cri.ImageCacheCopyStatusReady
+	case isoReady && diskReady:
+		// ready to copy
+		if err := ctrl.copyImageCache(ctx, logger, copySource, copyTarget); err != nil {
+			return nil, fmt.Errorf("error copying image cache: %w", err)
 		}
 
-		return optional.None[string](), false, fmt.Errorf("error getting volume status: %w", err)
+		copyStatus = cri.ImageCacheCopyStatusReady
+	default:
+		// waiting for copy preconditions
+		copyStatus = cri.ImageCacheCopyStatusPending
 	}
 
+	return &imageCacheVolumeStatus{
+		roots:      roots,
+		allReady:   allReady,
+		copyStatus: copyStatus,
+	}, nil
+}
+
+func (ctrl *ImageCacheConfigController) getImageCacheRoot(ctx context.Context, r controller.Reader, volumeStatus *block.VolumeStatus, mountReadyOnly bool) (optional.Optional[string], bool, error) {
```

Comment:
Why return an `Optional` if you also return a bool value?

Reply:
Ignore. Found my answer.

Reply:
because they are not same if you look closely.

It might be ready but missing path. 

Readiness is another dimension, it is ready if e.g. it was found to be missing, so the path is missing, but it's ready in the sense we don't need to wait for it